File: lib/minimal_tebd/tebd.py
"""Minimal TEBD engine.

Adapted from tenpy's toycode ``tenpy_toycodes/c_tebd.py``, with these changes:

- Real-time evolution by default: gates are ``exp(-1j * coeff * dt * H_bond)``.
- Supports Trotter orders 1, 2, 4 (4 = Forest--Ruth).
- The ``TEBDEngine`` class mimics tenpy's interface (``run()``,
  ``evolved_time``, ``trunc_err``) so downstream code written against tenpy
  can be swapped to this package with minimal changes.

The single hot kernel is :func:`update_bond` — this is the piece to
port to CUDA. Everything else is a thin driver.
"""

import numpy as np
from scipy.linalg import expm

# ---- Trotter schedules -----------------------------------------------------


def _trotter_substeps(order):
    """Return the list of ``(parity, dt_fraction)`` pairs for one time step.

    Each sub-step says: apply all gates of the given ``parity`` ('even' = bonds
    0, 2, 4, ... ; 'odd' = bonds 1, 3, 5, ...) for a fraction ``dt_fraction``
    of the full time step ``dt``.
    """
    if order == 1:
        return [('even', 1.0), ('odd', 1.0)]
    if order == 2:
        return [('even', 0.5), ('odd', 1.0), ('even', 0.5)]
    if order == 4:
        # Forest--Ruth 4th-order symplectic integrator, applied to the
        # splitting H = H_even + H_odd (within each group the bond Hamiltonians
        # commute, so we can exponentiate them independently).
        theta = 1.0 / (2.0 - 2.0 ** (1.0 / 3.0))
        a1 = theta / 2.0
        a2 = (1.0 - theta) / 2.0
        b1 = theta
        b2 = 1.0 - 2.0 * theta
        return [
            ('even', a1),
            ('odd', b1),
            ('even', a2),
            ('odd', b2),
            ('even', a2),
            ('odd', b1),
            ('even', a1),
        ]
    raise ValueError(f'unsupported order {order}; use 1, 2, or 4')


def _compute_gates(H_bonds, coeff):
    """Exponentiate each bond Hamiltonian: ``U_b = expm(coeff * H_b)``.

    Returned gates have shape ``(d, d, d, d)`` with index order
    ``(i_out, j_out, i_in, j_in)`` so they directly contract with ``theta2``.
    """
    gates = []
    for H in H_bonds:
        d = H.shape[0]
        H_mat = H.reshape(d * d, d * d)
        U = expm(coeff * H_mat)
        gates.append(U.reshape(d, d, d, d))
    return gates


# ---- SVD split + truncate (the inner loop of update_bond) ------------------


def split_truncate_theta(theta, chi_max, svd_min):
    """Split ``theta[vL, i, j, vR]`` by SVD, truncate, renormalise.

    Returns
    -------
    A : ndarray, shape ``(chi_L, d, chi_new)`` -- left-canonical on site i.
    S : ndarray, shape ``(chi_new,)`` -- Schmidt values on the new middle bond.
    B : ndarray, shape ``(chi_new, d, chi_R)`` -- right-canonical on site j.
    eps : float -- discarded weight (``1 - sum kept_S^2``), the bond truncation error.
    """
    chivL, dL, dR, chivR = theta.shape
    mat = theta.reshape(chivL * dL, dR * chivR)
    U, s, Vh = np.linalg.svd(mat, full_matrices=False)
    keep = min(chi_max, int(np.sum(s > svd_min)))
    if keep < 1:
        keep = 1  # never discard everything
    idx = np.argsort(s)[::-1][:keep]
    U = U[:, idx]
    s = s[idx]
    Vh = Vh[idx, :]
    norm = np.linalg.norm(s)
    eps = 1.0 - norm * norm
    S = s / norm
    A = U.reshape(chivL, dL, keep)
    B = Vh.reshape(keep, dR, chivR)
    return A, S, B, eps


# ---- the core kernel -------------------------------------------------------


def update_bond(psi, i, U_bond, chi_max, svd_min):
    """Apply a two-site gate on bond ``(i, i+1)`` in-place, truncate, return eps.

    This is the hot path: gate contraction + SVD + canonicalisation.
    Everything performance-critical happens here.
    """
    # Two-site wave function in mixed canonical form.
    theta = psi.get_theta2(i)  # vL i j vR
    # Contract with the gate on the two physical legs.
    # U_bond has legs (i', j', i, j); theta has (vL, i, j, vR).
    Utheta = np.tensordot(U_bond, theta, axes=([2, 3], [1, 2]))  # i' j' vL vR
    Utheta = np.transpose(Utheta, [2, 0, 1, 3])  # vL i' j' vR
    # SVD and truncate.
    A, S, B, eps = split_truncate_theta(Utheta, chi_max, svd_min)
    # Restore right-canonical B-form at site i:
    #   new_B[i] = diag(1/Ss[i]) @ A @ diag(S)
    # Clamp tiny Ss[i] to avoid blow-ups from stale small Schmidt values on the
    # neighbouring bond; this is the standard Vidal-form regularisation.
    Ss_i = psi.Ss[i]
    Ss_i_inv = np.where(Ss_i > 1e-14, 1.0 / Ss_i, 0.0)
    G_i = np.tensordot(np.diag(Ss_i_inv), A, axes=(1, 0))  # vL i vC
    psi.Bs[i] = np.tensordot(G_i, np.diag(S), axes=(2, 0))  # vL i vC
    psi.Ss[i + 1] = S
    psi.Bs[i + 1] = B
    return eps


# ---- accumulated truncation error -----------------------------------------


class TruncationError:
    """Running accumulator for truncation error, ``eps = 1 - prod_k (1 - eps_k)``.

    Mirrors (a small subset of) ``tenpy.algorithms.truncation.TruncationError``.
    """

    def __init__(self, eps=0.0):
        self.eps = float(eps)

    def add(self, eps_k):
        self.eps = self.eps + eps_k - self.eps * eps_k
        return self

    def __repr__(self):
        return f'TruncationError(eps={self.eps:.3e})'


# ---- TEBDEngine driver -----------------------------------------------------


class TEBDEngine:
    """TEBD time-evolution engine.

    Parameters
    ----------
    psi : SimpleMPS
        State (evolved in place).
    model : object with ``L``, ``d``, ``H_bonds`` attributes
        Typically a ``TFIChain`` or ``BondModel``.
    options : dict
        Recognised keys:

        - ``dt`` (float, default 0.1): time step.
        - ``order`` (int, default 2): Trotter order (1, 2, or 4).
        - ``N_steps`` (int, default 1): steps per ``run()`` call.
        - ``trunc_params`` (dict): ``chi_max`` (default 100) and
          ``svd_min`` (default 1e-12).
        - ``type_evo`` ('real' or 'imag', default 'real'): real-time or
          imaginary-time evolution. Imaginary time is non-unitary; use
          it to relax to a ground state.
    """

    def __init__(self, psi, model, options=None):
        options = dict(options or {})
        self.psi = psi
        self.model = model
        self.dt = float(options.get('dt', 0.1))
        self.order = int(options.get('order', 2))
        self.N_steps = int(options.get('N_steps', 1))
        trunc = options.get('trunc_params', {}) or {}
        self.chi_max = int(trunc.get('chi_max', 100))
        self.svd_min = float(trunc.get('svd_min', 1e-12))
        self.type_evo = options.get('type_evo', 'real')
        assert self.type_evo in ('real', 'imag')
        self.evolved_time = 0.0
        self.trunc_err = TruncationError()
        self._build_gate_cache()

    def _build_gate_cache(self):
        coeff_prefactor = -1j if self.type_evo == 'real' else -1.0
        self._schedule = _trotter_substeps(self.order)
        unique_fracs = sorted({frac for _, frac in self._schedule})
        self._gates = {
            frac: _compute_gates(self.model.H_bonds, coeff_prefactor * frac * self.dt)
            for frac in unique_fracs
        }

    def run(self):
        """Evolve the state by ``N_steps * dt``."""
        for _ in range(self.N_steps):
            self._one_trotter_step()
            self.evolved_time += self.dt

    def _one_trotter_step(self):
        L = self.model.L
        for parity, frac in self._schedule:
            gates = self._gates[frac]
            start = 0 if parity == 'even' else 1
            for b in range(start, L - 1, 2):
                eps = update_bond(self.psi, b, gates[b], self.chi_max, self.svd_min)
                self.trunc_err.add(eps)

File: src/core/tebd.cpp
#include "tebd.h"

#include <algorithm>
#include <cassert>
#include <cmath>
#include <cstring>
#include <set>
#include <vector>

#include "linalg.h"

// ===========================================================================
// Trotter schedules
// ===========================================================================

static int trotter_substeps(int order, int parities[], double fracs[]) {
  if (order == 1) {
    parities[0] = 0;
    fracs[0] = 1.0;
    parities[1] = 1;
    fracs[1] = 1.0;
    return 2;
  }
  if (order == 2) {
    parities[0] = 0;
    fracs[0] = 0.5;
    parities[1] = 1;
    fracs[1] = 1.0;
    parities[2] = 0;
    fracs[2] = 0.5;
    return 3;
  }
  if (order == 4) {
    double theta = 1.0 / (2.0 - std::pow(2.0, 1.0 / 3.0));
    double a1 = theta / 2.0;
    double a2 = (1.0 - theta) / 2.0;
    double b1 = theta;
    double b2 = 1.0 - 2.0 * theta;
    int p[] = {0, 1, 0, 1, 0, 1, 0};
    double f[] = {a1, b1, a2, b2, a2, b1, a1};
    for (int i = 0; i < 7; ++i) {
      parities[i] = p[i];
      fracs[i] = f[i];
    }
    return 7;
  }
  assert(false && "unsupported order; use 1, 2, or 4");
  return 0;
}

// ===========================================================================
// Compute gates: U_b = expm(coeff * H_bond_b)
// ===========================================================================

static void compute_gates(const BondModel& model, Cdouble coeff,
                          Tensor* gates) {
  int d = model.d;
  int d2 = d * d;
  std::vector<Cdouble> H_scaled(d2 * d2);
  std::vector<Cdouble> U_mat(d2 * d2);
  for (int b = 0; b < model.L - 1; ++b) {
    // Scale: H_scaled = coeff * H_bonds[b] (reshaped to d2 x d2).
    for (int i = 0; i < d2 * d2; ++i)
      H_scaled[i] = coeff * model.H_bonds[b].data[i];
    matrix_expm(H_scaled.data(), d2, U_mat.data());
    int sh[4] = {d, d, d, d};
    gates[b] = Tensor(4, sh);
    std::copy(U_mat.begin(), U_mat.end(), gates[b].data);
  }
}

// ===========================================================================
// SVD + truncation
// ===========================================================================

void svd_truncate(const Tensor& theta, int chi_max, double svd_min, Tensor& A,
                  double* S, Tensor& B, double& eps) {
  int chivL = theta.shape[0];
  int dL = theta.shape[1];
  int dR = theta.shape[2];
  int chivR = theta.shape[3];
  int m = chivL * dL;
  int n = dR * chivR;
  int mn = std::min(m, n);

  // Reshape theta to 2D (m x n) for SVD.
  // theta is already row-major (chivL, dL, dR, chivR), so reshape is free.
  std::vector<Cdouble> U_full(m * mn);
  std::vector<double> s_full(mn);
  std::vector<Cdouble> Vt_full(mn * n);

  // Handle m < n case: SVD requires m >= n.
  if (m >= n) {
    svd(theta.data, m, n, n, U_full.data(), s_full.data(), Vt_full.data());
  } else {
    // Transpose, SVD, then un-transpose.
    // A^T = V S U^T => A = U S V^T
    std::vector<Cdouble> At(n * m);
    for (int i = 0; i < m; ++i)
      for (int j = 0; j < n; ++j) At[j * m + i] = theta.data[i * n + j];
    std::vector<Cdouble> U_t(n * mn);
    std::vector<Cdouble> Vt_t(mn * m);
    svd(At.data(), n, m, m, U_t.data(), s_full.data(), Vt_t.data());
    // A^T = U_t S Vt_t, so A = Vt_t^T S U_t^T => U = Vt_t^T, Vt = U_t^T.
    U_full.resize(m * mn);
    Vt_full.resize(mn * n);
    for (int i = 0; i < m; ++i)
      for (int j = 0; j < mn; ++j)
        U_full[i * mn + j] = Vt_t[j * m + i];
    for (int i = 0; i < mn; ++i)
      for (int j = 0; j < n; ++j)
        Vt_full[i * n + j] = U_t[j * mn + i];
  }

  // Determine how many to keep.
  int keep = 0;
  for (int i = 0; i < mn; ++i) {
    if (s_full[i] > svd_min) ++keep;
  }
  keep = std::min(keep, chi_max);
  if (keep < 1) keep = 1;

  // s_full is already sorted descending by our SVD.
  // Compute truncation error.
  double norm_sq = 0.0;
  for (int i = 0; i < mn; ++i) norm_sq += s_full[i] * s_full[i];
  double kept_sq = 0.0;
  for (int i = 0; i < keep; ++i) kept_sq += s_full[i] * s_full[i];
  eps = 1.0 - kept_sq;  // before renormalization

  // Renormalize.
  double norm = std::sqrt(kept_sq);
  for (int i = 0; i < keep; ++i) S[i] = s_full[i] / norm;

  // A: shape (chivL, dL, keep)
  int shA[3] = {chivL, dL, keep};
  A = Tensor(3, shA);
  for (int i = 0; i < m; ++i)
    for (int j = 0; j < keep; ++j) A.data[i * keep + j] = U_full[i * mn + j];

  // B: shape (keep, dR, chivR)
  int shB[3] = {keep, dR, chivR};
  B = Tensor(3, shB);
  for (int i = 0; i < keep; ++i)
    for (int j = 0; j < n; ++j) B.data[i * n + j] = Vt_full[i * n + j];
}

// ===========================================================================
// update_bond
// ===========================================================================

double update_bond(MPS* psi, int site, const Tensor& gate, int chi_max,
                   double svd_min) {
  int d = psi->d;

  // theta2: (chi_L, d, d, chi_R)
  Tensor theta = psi->theta2(site);
  int chivL = theta.shape[0];
  int chivR = theta.shape[3];

  // Apply gate: U(i',j',i,j) * theta(vL,i,j,vR) -> Utheta(i',j',vL,vR)
  // Then transpose to (vL, i', j', vR).
  int sh_out[4] = {chivL, d, d, chivR};
  Tensor Utheta(4, sh_out);
  for (int vL = 0; vL < chivL; ++vL) {
    for (int ip = 0; ip < d; ++ip) {
      for (int jp = 0; jp < d; ++jp) {
        for (int vR = 0; vR < chivR; ++vR) {
          Cdouble s = 0.0;
          for (int i = 0; i < d; ++i)
            for (int j = 0; j < d; ++j)
              s += gate(ip, jp, i, j) * theta(vL, i, j, vR);
          Utheta(vL, ip, jp, vR) = s;
        }
      }
    }
  }

  // SVD and truncate.
  Tensor A_out, B_out;
  std::vector<double> S_buf(chi_max > 0 ? chi_max : 1);
  double eps;
  svd_truncate(Utheta, chi_max, svd_min, A_out, S_buf.data(), B_out, eps);

  int keep = A_out.shape[2];

  // Restore right-canonical B-form at site i.
  // new_Bs[i] = diag(1/Ss[i]) @ A @ diag(S)
  int chi_L = psi->Bs[site].shape[0];
  int sh_Bi[3] = {chi_L, d, keep};
  Tensor new_Bi(3, sh_Bi, psi->Bs[site].capacity);
  for (int a = 0; a < chi_L; ++a) {
    double ss_inv =
        (psi->Ss[site][a] > 1e-14) ? 1.0 / psi->Ss[site][a] : 0.0;
    for (int p = 0; p < d; ++p) {
      for (int b = 0; b < keep; ++b) {
        new_Bi(a, p, b) = ss_inv * A_out(a, p, b) * S_buf[b];
      }
    }
  }

  psi->Bs[site] = std::move(new_Bi);

  // Update Ss[site+1] and Bs[site+1].
  delete[] psi->Ss[site + 1];
  psi->Ss[site + 1] = new double[psi->chi_max];
  std::fill(psi->Ss[site + 1], psi->Ss[site + 1] + psi->chi_max, 0.0);
  for (int i = 0; i < keep; ++i) psi->Ss[site + 1][i] = S_buf[i];
  psi->chi[site + 1] = keep;

  // B_out already has correct shape (keep, d, chivR). Set with capacity.
  int shB[3] = {keep, d, chivR};
  Tensor new_Bj(3, shB, psi->Bs[site + 1].capacity);
  std::copy(B_out.data, B_out.data + B_out.size, new_Bj.data);
  psi->Bs[site + 1] = std::move(new_Bj);

  return eps;
}

// ===========================================================================
// TEBDEngine
// ===========================================================================

TEBDEngine::TEBDEngine(MPS* psi, BondModel* model, double dt, int order,
                       int N_steps, int chi_max, double svd_min, int type_evo)
    : psi(psi),
      model(model),
      dt(dt),
      order(order),
      N_steps(N_steps),
      chi_max(chi_max),
      svd_min(svd_min),
      type_evo(type_evo),
      evolved_time(0.0),
      trunc_err_eps(0.0),
      n_schedule(0),
      schedule(nullptr),
      n_unique_fracs(0),
      unique_fracs(nullptr),
      gate_sets(nullptr) {
  assert(type_evo == 0 || type_evo == 1);
  build_gate_cache();
}

TEBDEngine::~TEBDEngine() {
  delete[] schedule;
  delete[] unique_fracs;
  if (gate_sets) {
    for (int i = 0; i < n_unique_fracs; ++i) delete[] gate_sets[i];
    delete[] gate_sets;
  }
}

void TEBDEngine::build_gate_cache() {
  int parities[7];
  double fracs[7];
  n_schedule = trotter_substeps(order, parities, fracs);
  schedule = new TrotterStep[n_schedule];

  // Collect unique fractions.
  std::set<double> frac_set;
  for (int i = 0; i < n_schedule; ++i) frac_set.insert(fracs[i]);
  n_unique_fracs = frac_set.size();
  unique_fracs = new double[n_unique_fracs];
  int idx = 0;
  for (double f : frac_set) unique_fracs[idx++] = f;

  // Map schedule entries to frac indices.
  for (int i = 0; i < n_schedule; ++i) {
    schedule[i].parity = parities[i];
    for (int k = 0; k < n_unique_fracs; ++k) {
      if (std::abs(unique_fracs[k] - fracs[i]) < 1e-15) {
        schedule[i].frac_idx = k;
        break;
      }
    }
  }

  // Compute gate sets.
  Cdouble coeff_prefactor = (type_evo == 0) ? Cdouble(0, -1) : Cdouble(-1, 0);
  gate_sets = new Tensor*[n_unique_fracs];
  for (int k = 0; k < n_unique_fracs; ++k) {
    gate_sets[k] = new Tensor[model->L - 1];
    Cdouble coeff = coeff_prefactor * unique_fracs[k] * dt;
    compute_gates(*model, coeff, gate_sets[k]);
  }
}

void TEBDEngine::run() {
  for (int s = 0; s < N_steps; ++s) {
    step();
    evolved_time += dt;
  }
}

void TEBDEngine::step() {
  int L = model->L;
  for (int s = 0; s < n_schedule; ++s) {
    int start = schedule[s].parity;  // 0 = even, 1 = odd
    Tensor* gates = gate_sets[schedule[s].frac_idx];
    for (int b = start; b < L - 1; b += 2) {
      double eps = update_bond(psi, b, gates[b], chi_max, svd_min);
      trunc_err_eps = trunc_err_eps + eps - trunc_err_eps * eps;
    }
  }
}
